drivers/gpio/gpio_ite_it8xxx2.c

```diff
@@ -0,0 +1,474 @@
+{
+	uint8_t regv = ite_read(config->reg_addr, 1);
+
+	return !!(regv & BIT(bit));
+}
+
+static inline void set_port(const struct gpio_ite_cfg *config, uint8_t value)
+{
+	ite_write(config->reg_addr, 1, value);
+}
+
+static inline uint8_t get_port(const struct gpio_ite_cfg *config)
+{
+	uint8_t regv = ite_read(config->reg_addr, 1);
+
+	return (regv & BIT(config->gpios_pin));
+}
+
+/* Driver functions */
+
+static int gpio_ite_configure(struct device *dev,
+			      gpio_pin_t pin, gpio_flags_t flags)
+{
+	const struct gpio_ite_cfg *gpio_config = DEV_GPIO_CFG(dev);
+	unsigned int gpcr_offset = 0x10;
+	uint32_t gpcr_reg;
+	uint32_t gpcr_reg_addr;
+
+	gpcr_reg = ((gpio_config->reg_addr & 0xff)  - 1) * 8 + gpcr_offset;
+	gpcr_reg_addr =  gpcr_reg | (gpio_config->reg_addr & 0xffffff00);
+
+
+	if ((flags & GPIO_OUTPUT) && (flags & GPIO_INPUT)) {
+		/* Pin cannot be configured as input and output */
+		return -ENOTSUP;
+	} else if (!(flags & (GPIO_INPUT | GPIO_OUTPUT))) {
+		/* Pin has to be configuread as input or output */
+		return -ENOTSUP;
+	}
+
+
+	if (flags & GPIO_OUTPUT) {
+
+		ite_write(gpcr_reg_addr + pin, 1, 0x40);
+		if (flags & GPIO_OUTPUT_INIT_HIGH) {
+			set_bit(gpio_config, pin, GPIO_HIGH);
+		} else if (flags & GPIO_OUTPUT_INIT_LOW) {
+			set_bit(gpio_config, pin, GPIO_LOW);
+		}
+	} else {
+
+		ite_write(gpcr_reg_addr + pin, 1, 0x80);
+		if ((flags & GPIO_PULL_DOWN) || (flags & GPIO_PULL_UP)) {
+			return -ENOTSUP;
+		}
+
+	}
+
+	return 0;
+}
+static int gpio_ite_port_get_raw(struct device *dev, gpio_port_value_t *value)
+{
+	const struct gpio_ite_cfg *gpio_config = DEV_GPIO_CFG(dev);
+
+	*value = ite_read(gpio_config->reg_addr, 1);
+	return 0;
+}
+
+static int gpio_ite_port_set_masked_raw(struct device *dev,
+					gpio_port_pins_t mask,
+					gpio_port_value_t value)
+{
+	const struct gpio_ite_cfg *gpio_config = DEV_GPIO_CFG(dev);
+	uint32_t port_val;
+
+	port_val = get_port(gpio_config);
+	port_val = (port_val & ~mask) | (value & mask);
+	set_port(gpio_config, port_val);
+	return 0;
+}
+
+static int gpio_ite_port_set_bits_raw(struct device *dev,
+				      gpio_port_pins_t pins)
+{
+	const struct gpio_ite_cfg *gpio_config = DEV_GPIO_CFG(dev);
+	uint32_t port_val;
+
+	port_val = get_port(gpio_config);
+	port_val |= pins;
+	set_port(gpio_config, port_val);
+	return 0;
+}
+
+static int gpio_ite_port_clear_bits_raw(struct device *dev,
+					gpio_port_pins_t pins)
+{
+	const struct gpio_ite_cfg *gpio_config = DEV_GPIO_CFG(dev);
+	uint32_t port_val;
+
+	port_val = get_port(gpio_config);
+	port_val &= ~pins;
+	set_port(gpio_config, port_val);
+	return 0;
+}
+
+static int gpio_ite_port_toggle_bits(struct device *dev,
+				     gpio_port_pins_t pins)
+{
+	const struct gpio_ite_cfg *gpio_config = DEV_GPIO_CFG(dev);
+	uint32_t port_val;
+
+	port_val = get_port(gpio_config);
+	port_val ^= pins;
+	set_port(gpio_config, port_val);
+	return 0;
+}
+
+static void gpio_ite_isr(void *arg)
+{
+	int i;
+	struct device *dev = arg;
```

Comment:
Cast void* to target type

Reply:
why would that be necessary? the C compiler lets you assign void * to other pointer types without complaint

Reply:
Sorry, you're right. It's C++ needs the explicit cast.